### ISSUE TITLE
Variadic aggregations with nullable parameters might use garbage data

### ISSUE THREAD

**Issue by Algunenano** [ClickHouse/ClickHouse#41528]
> You have to provide the following information whenever possible.

When the return value of an aggregation cannot be nullable, for example a tuple, the aggregation function might use the undelying / garbage data of a NULL value.

Example on `22.8.5.29`:

Data:
```sql
SELECT
    argMax((n, n), n) AS t,
    toTypeName(t)
FROM
(
    SELECT if((number % 3) = 0, NULL, number) AS n
    FROM numbers(10)
)

Query id: b74a4de2-8738-4434-947f-6945f3c22762

┌─t───────────┬─toTypeName(argMax(tuple(n, n), n))────────┐
│ (NULL,NULL) │ Tuple(Nullable(UInt64), Nullable(UInt64)) │
└─────────────┴───────────────────────────────────────────┘
```

Here I was expecting to get `(8, 8)`. Why didn't we? Let's see the data:

```
SELECT
    (n, n),
    n,
    assumeNotNull(n)
FROM
(
    SELECT if((number % 3) = 0, NULL, number) AS n
    FROM numbers(10)
)

Query id: a3fad631-5c79-47c4-9a63-89258cf17848

┌─tuple(n, n)─┬────n─┬─assumeNotNull(n)─┐
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                0 │
│ (1,1)       │    1 │                1 │
│ (2,2)       │    2 │                2 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                3 │
│ (4,4)       │    4 │                4 │
│ (5,5)       │    5 │                5 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                6 │
│ (7,7)       │    7 │                7 │
│ (8,8)       │    8 │                8 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                9 │
└─────────────┴──────┴──────────────────┘
```

The last value of n is a NULLed 9, which is considered bigger than a non-NULLed 8 (because the null qualifiers are removed and ignored).

OTOH, if we force a cast the value will be default'ed:

```
SELECT
    argMax((n, n), n) AS t,
    toTypeName(t)
FROM
(
    SELECT if((number % 3) = 0, NULL, CAST(number, 'Int32')) AS n
    FROM numbers(10)
)

Query id: 52e778ce-4530-4c58-a0d9-2adcaa9d5f41

┌─t─────┬─toTypeName(argMax(tuple(n, n), n))──────┐
│ (8,8) │ Tuple(Nullable(Int32), Nullable(Int32)) │
└───────┴─────────────────────────────────────────┘

1 row in set. Elapsed: 0.002 sec. 
```

This is only because the NULLed 9 was transformed into a NULLed 0 so the result seems fine:

```
SELECT
    (n, n),
    n,
    assumeNotNull(n)
FROM
(
    SELECT if((number % 3) = 0, NULL, CAST(number, 'Int32')) AS n
    FROM numbers(10)
)

Query id: da7dfc57-a5bb-4f13-a8e7-d52c9b0f9aa5

┌─tuple(n, n)─┬────n─┬─assumeNotNull(n)─┐
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                0 │
│ (1,1)       │    1 │                1 │
│ (2,2)       │    2 │                2 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                0 │
│ (4,4)       │    4 │                4 │
│ (5,5)       │    5 │                5 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                0 │
│ (7,7)       │    7 │                7 │
│ (8,8)       │    8 │                8 │
│ (NULL,NULL) │ ᴺᵁᴸᴸ │                0 │
└─────────────┴──────┴──────────────────┘

10 rows in set. Elapsed: 0.001 sec.
```

The problem now is that the 0 can then cause issues with the min :D: 

```
Select argMin((n, n), n) t, toTypeName(t) FROM (Select if(number % 3 == 0, NULL, number::Int32) as n from numbers(5, 10))

SELECT
    argMin((n, n), n) AS t,
    toTypeName(t)
FROM
(
    SELECT if((number % 3) = 0, NULL, CAST(number, 'Int32')) AS n
    FROM numbers(5, 10)
)

Query id: 71f7283d-c7db-48cb-91f9-2d1f6a34d790

┌─t───────────┬─toTypeName(argMin(tuple(n, n), n))──────┐
│ (NULL,NULL) │ Tuple(Nullable(Int32), Nullable(Int32)) │
└─────────────┴─────────────────────────────────────────┘

1 row in set. Elapsed: 0.002 sec. 
```

And removing the cast now returns the good value:

```
SELECT
    argMin((n, n), n) AS t,
    toTypeName(t)
FROM
(
    SELECT if((number % 3) = 0, NULL, number) AS n
    FROM numbers(5, 10)
)

Query id: c1292787-23d5-4402-bbdf-5b51e6b8e028

┌─t─────┬─toTypeName(argMin(tuple(n, n), n))────────┐
│ (5,5) │ Tuple(Nullable(UInt64), Nullable(UInt64)) │
└───────┴───────────────────────────────────────────┘

1 row in set. Elapsed: 0.002 sec. 
```

Basically I'm expecting:
* NULL values, in any column, will **always be ignored**; which is similar to what unary aggregation do (or binary+ aggregations that can return NULL).
* If no value was added to the aggregation and the return value is not nullable, then it should return the default value. In this case `(NULL,NULL)::Tuple(Nullable(UInt64), Nullable(UInt64))`

Seems like a straightforward change (but something that could be considered breaking change). Not sure if I'm missing something. Thoughts?

**Comment by Algunenano** [comment]
Looking at the code it seems that this option to not ignore nulls was introduced for some specific functions and it might be possible that it crept to be the default behaviour by mistake.

The 3 aggregation functions using this option manually are:
* windowFunnel
* intervalLengthSum
* sequenceNextNode 

It doesn't make sense for argMax to receive an add() with random crap if one of the columns is NULL, that's for sure, but I need to check if any of those other 3 functions really are expected to work correctly with "random" data. I doubt they will be doing the correct thing, but you never know.

**Comment by Algunenano** [comment]
* windowFunnel: Does have tests around NULL values but they seem to assume that the underlying value of NULL is equal to 0/default in all situations, when it's not.
* intervalLengthSum: No idea. Does not have tests with NULLs so it's hard to say what's supposed to be the expected behaviour 
* sequenceNextNode: It's ignoring NULLs already, so it wouldn't need a custom NULL adapter.

So to me it's clear that this is buggy and shouldn't exists as is. Not only that, but it goes against the documentation (https://clickhouse.com/docs/en/sql-reference/aggregate-functions/#null-processing). I think the best way forward is to remove this behaviour and always skip nulls, recommend people using NULLs that want to use some other value to use COALESCE of IF or whatever other method they want and set a value that matches their desired value and forget this ever existed. Obviously I'd call this a breaking change, although I really believe it's a pretty bad bug.

**Comment by UnamedRus** [comment]
IMO, there were quite a few people, who relied on that behavior, that tuple will prevent NULL from being ignored...

**Comment by Algunenano** [comment]
> IMO, there were quite a few people, who relied on that behavior, that tuple will prevent NULL from being ignored...

Isn't that still the case?

```
SELECT
    argMax(tuple(t), n).1,
    argMin(tuple(t), n).1
FROM
(
    SELECT
        NULL AS t,
        5 AS n
    UNION ALL
    SELECT
        1 AS t,
        4 AS n
)

Query id: eb4863f5-7967-418d-84a0-a33a2eee51ac

┌─tupleElement(argMax(tuple(t), n), 1)─┬─tupleElement(argMin(tuple(t), n), 1)─┐
│                                 ᴺᵁᴸᴸ │                                    1 │
└──────────────────────────────────────┴──────────────────────────────────────┘
```

**Comment by UnamedRus** [comment]
Yeah, you are right.

Did not interpreter examples correctly.
